collections/src/main/scala/strawman/collection/immutable/IntMap.scala

```diff
@@ -0,0 +1,465 @@
+    case IntMap.Bin(_, _, left, right) => { left.foreachValue(f); right.foreachValue(f) }
+    case IntMap.Tip(_, value) => f(value)
+    case IntMap.Nil =>
+  }
+
+  override def className = "IntMap"
+
+  override def isEmpty = this == IntMap.Nil
+
+  override def filter(f: ((Int, T)) => Boolean): IntMap[T] = this match {
+    case IntMap.Bin(prefix, mask, left, right) => {
+      val (newleft, newright) = (left.filter(f), right.filter(f))
+      if ((left eq newleft) && (right eq newright)) this
+      else bin(prefix, mask, newleft, newright)
+    }
+    case IntMap.Tip(key, value) =>
+      if (f((key, value))) this
+      else IntMap.Nil
+    case IntMap.Nil => IntMap.Nil
+  }
+
+  def transform[S](f: (Int, T) => S): IntMap[S] = this match {
+    case b@IntMap.Bin(prefix, mask, left, right) => b.bin(left.transform(f), right.transform(f))
+    case t@IntMap.Tip(key, value) => t.withValue(f(key, value))
+    case IntMap.Nil => IntMap.Nil
+  }
+
+  final override def size: Int = this match {
+    case IntMap.Nil => 0
+    case IntMap.Tip(_, _) => 1
+    case IntMap.Bin(_, _, left, right) => left.size + right.size
+  }
+
+  final def get(key: Int): Option[T] = this match {
+    case IntMap.Bin(prefix, mask, left, right) => if (zero(key, mask)) left.get(key) else right.get(key)
+    case IntMap.Tip(key2, value) => if (key == key2) Some(value) else None
+    case IntMap.Nil => None
+  }
+
+  final override def getOrElse[S >: T](key: Int, default: => S): S = this match {
+    case IntMap.Nil => default
+    case IntMap.Tip(key2, value) => if (key == key2) value else default
+    case IntMap.Bin(prefix, mask, left, right) =>
+      if (zero(key, mask)) left.getOrElse(key, default) else right.getOrElse(key, default)
+  }
+
+  final override def apply(key: Int): T = this match {
+    case IntMap.Bin(prefix, mask, left, right) => if (zero(key, mask)) left(key) else right(key)
+    case IntMap.Tip(key2, value) => if (key == key2) value else throw new IllegalArgumentException("Key not found")
+    case IntMap.Nil => throw new IllegalArgumentException("key not found")
+  }
+
+  override def + [S >: T] (kv: (Int, S)): IntMap[S] = updated(kv._1, kv._2)
+
+  override def updated[S >: T](key: Int, value: S): IntMap[S] = this match {
+    case IntMap.Bin(prefix, mask, left, right) =>
+      if (!hasMatch(key, prefix, mask)) join(key, IntMap.Tip(key, value), prefix, this)
+      else if (zero(key, mask)) IntMap.Bin(prefix, mask, left.updated(key, value), right)
+      else IntMap.Bin(prefix, mask, left, right.updated(key, value))
+    case IntMap.Tip(key2, value2) =>
+      if (key == key2) IntMap.Tip(key, value)
+      else join(key, IntMap.Tip(key, value), key2, this)
+    case IntMap.Nil => IntMap.Tip(key, value)
+  }
+
+  def map[V2](f: ((Int, T)) => (Int, V2)): IntMap[V2] = intMapFromIterable(View.Map(toIterable, f))
+
+  def flatMap[V2](f: ((Int, T)) => IterableOnce[(Int, V2)]): IntMap[V2] = intMapFromIterable(View.FlatMap(toIterable, f))
```

Comment:
For the sake of completeness you might also want to add an overload for `concat`. And you probably want to refine `withFilter` as well.

Reply:
I'll add `concat`. I was unable to implement `withFilter` without triggering the same compiler bug I ran into earlier (when `WithFilter` was still a trait, IIRC, which is what it has to be in order to get the types right in these cases). I'll try to get back to this at some point and remove the path-dependent `WithFilter` types.